### ISSUE TITLE
protoc-gen-go: reduce initialization cost of generated code

### ISSUE THREAD

**Issue by JimRoewe** [golang/protobuf#862]
The init of generated protobuf code allocates ~225KB for our use case and will only continue to grow over time as continue to use protobuf. We run on an embedded platform with limited memory. We have 20 processes running, so this adds up to 4.5MB. 80KB of this is in http2/hpack.init which was recently improved to be lazy loading.

In general, I would really like if protobuf could do more lazy loading and encapsulate more data inside of functions instead of in global variables. I used pprof on our process, but I can't clearly see where all of the memory is going in our protobuf.init. 40KB is used by proto.RegisterType, but there's almost 100KB that isn't explicitly listed. I expect that all of the global variables used for fileDescriptor data and the enum maps are the cause.

I don't know enough about the RegisterType data to know how easy that would be to lazy load, but the fileDescriptor data should be easy to encapsulate in a function. The enum maps should also be easy to change to be inside of a function, but I understand that would be a code compatibility breaking change. Maybe having a code generation flag to put the enum information into functions would be possible?

I'm open to other ideas about how to reduce memory use. It would be nice to find a way to  reduce memory to only what's actually used in a process. Given that core golang made improvements in this recently, I hope we can find some improvements in protobuf too.

**Comment by dsnet** [comment]
> In general, I would really like it if protobuf could do more lazy loading and encapsulate more data inside of functions instead of in global variables

This is overall challenging since the protobuf language itself is designed around the assumption of a global registry to properly support extensions and [Any](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto) messages. Thus, some amount of initialization logic is always going to be needed for proper protobuf semantics.

> The enum maps should also be easy..., but I understand that would be a code compatibility breaking change

The latest version of the generator actually does not directly use the enum maps at all. However, golang/go#2559 prevents the compiler from currently eliminating unused maps as dead code. I'd like to see a compiler fix for this before providing a flag to drop generation of them.

Overall, there is much room to reduce the initialization cost.

**Comment by JimRoewe** [comment]
I understand that a global registry is needed. It would be nice if it could be structured to have functions or interfaces registered instead of raw data or use some form of hierarchical lazy-ish registration.
As a starting point idea, a pb.go file could have a single function that it passes for registration and that function could be called lazily, later to do all the registration. Or the registration functions could be set up to only register top level messages and when data is needed for a submessage, call sub message registration from the relevant top level message.

I know there's a lot of functionality around "Any" and extensions. If a lot of the code or memory cost is only for those areas, having an option to exclude that functionality would be nice.

The maps made by the generator are public (we currently use them), so removing them would probably break compatibility (that's what I was trying to mention earlier). But I'd be happy to update our code to use functions for equivalent lookups instead of using the maps if there was a way to make the generator provide enum lookup functions instead of maps.

Can you explain what you mean by you'd rather see a compiler fix for eliminating unused maps as dead code compared to a flag to drop generation of the maps?
Is your point that you'd rather not have an option to prevent generation of the maps and instead have the memory consumption solved by the compiler? If I understood that correctly, I would still want to have the option to not provide maps because they add a non-obvious cost. We have many developers working on our code base and I'd rather prevent them from using the maps because we turn them off then have to educate all of the developers to not use the maps (because using the maps adds memory overhead.)

Lastly, you said "Overall, there is much room to reduce the initialization cost." Was that a typo? Are you saying there IS room to reduce initialization cost? Or did you mean to say there isn't room to reduce cost? If there is room to reduce cost, can you mention some possibilities?

**Comment by dsnet** [comment]
The release of the `google.golang.org/protobuf` module has completely changed the implementation of initialization. I suspect it is faster, but it has not been benchmarked.

> It would be nice if it could be structured to have functions or interfaces registered instead of raw data or use some form of hierarchical lazy-ish registration.

I don't know how that is feasible given that we need to check for conflicts and other issues that requires the data be in a structured form.

> I would still want to have the option to not provide maps because they add a non-obvious cost.

An option just to avoid generation of maps is unlikely to happen. This project tries hard to keep the amount of generation knobs and whistles down. A knob just to disable the generated maps for enums doesn't worth it. I still stand by my original point of waiting for the Go compiler elide dead maps.

Closing this issue as there doesn't seem to be anything directly actionable to do here. If there are specific bottlenecks that have been identified (and there's something to be done about it), they can be a separate and more targeted issue. As for lazy registration, we've already made it as lazy as can be without breaking semantics. As for the generated maps, let the compiler fix problem of dead map elimination (golang/go#2559).
